### PR TITLE
Tei id extractor handles new files

### PR DESCRIPTION
**Pull request by alicefuzier**
For reference as to what the TEI Id extractor is https://github.com/wellcomecollection/docs/pull/62
This PR makes the TEI id extractor:
* receive a message about a path in the tei repo that has changed
* read the tei file from github
* get the id from the file
* store the file in s3
* send a message with the location in s3 of the file and the id

It doesn't store the path and the id anywhere yet, and it doesn't handle files deleted or moved. That will be added in a subsequent pr
Part of https://github.com/wellcomecollection/catalogue-pipeline/issues/1536

### CODE REVIEW

**Comment by alexwlchan** [comment]
I'm going to review this now.

**Review by alexwlchan — CHANGES_REQUESTED** [review]
This mostly looks good, and was less of a tangle to review than I was expecting, but I'd like more in the way of signposting and comments for future readers to understand why it's implemented this way. Specific suggestions and comments inline.

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/Main.scala
Comment (alicefuzier) [reply tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/Main.scala:24]: This is fixed in #1687 .  I haven't bothered to do it properly here cause this version is not meant to be deployed ( In fact the deployed version is from #1687 )

Comment (alexwlchan) [reply tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/Main.scala:24]: Gotcha.

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobReader.scala
Comment (alicefuzier) [reply tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobReader.scala:27]: Actually I don't think we do. The github documentation recommends - but doesn't require - the Accept header with the api version, so that things don't shift under your feet. EVerything else was me trying to understand why I was getting 403 and it turned out it was rate limiting.
I will delete the ones we don't need and comment the one we do need

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobReader.scala
Comment (alicefuzier) [reply tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobReader.scala:31]: Thanks sounds sensible. I will keep using the AkkaHttpClient for testing too though, cause I don't trust my understanding of the returned HttpResponse. I'd rather leave it to Wiremock to record and send back without me having to create a response from that and assume they are the same.

One side question while I'm adding it: why does [AkkaHttpClient](https://github.com/wellcomecollection/scala-libs/blob/main/http/src/main/scala/weco/http/client/AkkaHttpClient.scala) want both an actor system and an execution  context in the constructor? can it not derive the execution context from the actor system?

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobReader.scala
Comment (alicefuzier) [reply tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobReader.scala:46]: why are the other ones different? maybe I'm not understanding how it works, but I though the circe support was just adding marshallers, not encodes/decoders, and if you have in scope your own decoder/encoders it will use those, or so I though

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/TeiIdExtractorWorkerService.scala
Comment (alicefuzier) [reply tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/TeiIdExtractorWorkerService.scala:34]: I can put a link to the tei repo, i think looking at that is pretty self explanatory.
The only approach that I can think of to deduce programmatically is going via a blacklist instead of a whitelist approach like here, but I think it might be a bit more fiddly? Also, I don't think new languages are added very often. 
This field could be loaded from config and everything could be moved in terraform which at least skips the pr/release/deploy bit. I don't think it's very useful doing that right now though.
I think this is something to keep in mind but IMO this is good enough for now

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobReader.scala
Comment (alexwlchan) [reply tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobReader.scala:31]: > I will keep using the AkkaHttpClient for testing too though, cause I don't trust my understanding of the returned HttpResponse. I'd rather leave it to Wiremock to record and send back without me having to create a response from that and assume they are the same.

👍

> why does AkkaHttpClient want both an actor system and an execution context in the constructor? can it not derive the execution context from the actor system?

Good spot! That looks like a leftover from an older version of the code, I'll fix that now.

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobReader.scala
Comment (alicefuzier) [reply tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobReader.scala:46]: Unless there's a strong opposition I think I'm going to keep the marshalling library. It still allows to pass an implicit decoder instead of an implicit unmarshaller, but it can create the unmarshaller from the decoder implicitly. Also, I like the control with the level of error details.
I'm going to change the method to have an implicit decoder instead of an implicit unmarshaller though

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/TeiIdExtractorWorkerService.scala
Comment (alicefuzier) [reply tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/TeiIdExtractorWorkerService.scala:23]: No reason..It allows for different timezones while Instant is always UTC but actually that's good and ZonedDateTime is an overkill. WIll change

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/IdExtractor.scala
Comment (alicefuzier) [reply tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/IdExtractor.scala:16]: I have tested this and it fails if there's more than one xml:id attribute in the same node cause it's not valid xml. It also doesn't pick up other xml:id attributes in other nodes. So the code is fine. I've committed the extra tests

**Comment by alicefuzier** [comment]
I think I've addressed all your comments @alexwlchan

File: common/flows/src/main/scala/weco/flows/FlowOps.scala
Comment (alexwlchan) [inline common/flows/src/main/scala/weco/flows/FlowOps.scala:59]: It might be helpful to explain what "merge" means in this context – you get a new flow with all the outputs from both flows, in any order.

e.g. if `a` produced `(1, 2, 3, …)` and `b` produced `(-1, -2, -3, …)`, you might get `(1, -1, 2, 3, 4, -2, 5, -3)`.

Or have I misunderstood?

File: common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala
Comment (alexwlchan) [inline common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala:27]: Why has `ec` become a `val`?

File: pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/services/MergerWorkerService.scala
Comment (alexwlchan) [inline pipeline/merger/src/main/scala/uk/ac/wellcome/platform/merger/services/MergerWorkerService.scala:36]: Ditto above.

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala
Comment (alexwlchan) [inline tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala:23]: 👍

Super, thanks.

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala
Comment (alexwlchan) [inline tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala:64]: nit: 
```suggestion
              s"The GitHub api returned an error: ${code.value} for url ${request.uri} with headers ${response.headers}: $entity (couldn't decode entity as string)"))
```

This will make it more obvious which path you went down in the logs, and if memory serves the string-ified representation of an `HttpEntity` isn't particularly helpful.

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala
Comment (alexwlchan) [inline tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala:70]: ✨

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/IdExtractor.scala
Comment (alexwlchan) [inline tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/IdExtractor.scala:10]: Super. ✨

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/IdExtractor.scala
Comment (alexwlchan) [inline tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/IdExtractor.scala:22]:
```suggestion
        Failure(new RuntimeException(s"Unable to extract ID from XML at $uri", th))
```

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/TeiIdExtractorWorkerService.scala
Comment (alexwlchan) [inline tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/TeiIdExtractorWorkerService.scala:21]: Explaining that these come from the Lambda is helpful. 👍

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/TeiIdExtractorWorkerService.scala
Comment (alexwlchan) [inline tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/TeiIdExtractorWorkerService.scala:78]: Break this up to make it more readable:

```suggestion
            location = S3ObjectLocation(
              bucket = config.bucket,
              key = s"tei_files/$id/${message.timeModified.getEpochSecond}.xml")    
            )
            stored <- store.put(location, blobContent) match {
              case Right(stored) => Future.successful(stored)
              case Failure(error) => Future.failed(error.e)
            }
```

File: tei_adapter/tei_id_extractor/src/test/scala/weco/catalogue/tei/id_extractor/IdExtractorTest.scala
Comment (alexwlchan) [inline tei_adapter/tei_id_extractor/src/test/scala/weco/catalogue/tei/id_extractor/IdExtractorTest.scala:55]: Nice test. ✨

**Review by alexwlchan — APPROVED** [review]
Looking good, thanks for addressing my comments. ✨ 

A bunch more minor suggestions, but nothing that needs another review from me.

File: common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala
Comment (alicefuzier) [reply common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala:27]: I'm not 100% sure.. After moving out the broadcastAndMerge thing it complained that it didn't have access to the ExecutionContext 🤷‍♀️

File: common/flows/src/main/scala/weco/flows/FlowOps.scala
Comment (alicefuzier) [reply common/flows/src/main/scala/weco/flows/FlowOps.scala:59]: yes, I will add it

File: common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala
Comment (kenoir) [reply common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala:27]: Does the `logging` trait expect this perhaps?

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala
Comment (kenoir) [inline tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala:16]: Has this got any TEI logic in it? If not would it work to change the namespace here so it's not part of the id_extractor?

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala
Comment (kenoir) [inline tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala:20]: I wonder if this would benefit from the HttpClient pattern @alexwlchan abstracted recently (https://github.com/wellcomecollection/scala-libs/blob/main/http/src/main/scala/weco/http/client/HttpClient.scala) for reducing WireMock interactions?

**Review by kenoir — APPROVED** [review]

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala
Comment (alicefuzier) [reply tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala:20]: There's a conversation with @alexwlchan above about this. I don't want to bypass wiremock cause that requires me understanding properly the response and converting in an HttpResponse object accurately, and I don't trust myself on this, I'd rather delegate it to wiremock

File: tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala
Comment (alicefuzier) [reply tei_adapter/tei_id_extractor/src/main/scala/weco/catalogue/tei/id_extractor/GitHubBlobContentReader.scala:16]: No tei. It could technically live elsewhere although I'm not sure it's worth it right now

File: common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala
Comment (alicefuzier) [reply common/pipeline_storage/src/main/scala/uk/ac/wellcome/pipeline_storage/PipelineStorageStream.scala:27]: I don't think it's the logging, it shouldn't need the ExecutionContext. IIRC, in case classes any param in the class is automatically converted to a val so it's available to any method in the class. In normal classes, they're not so, unless you declare them val, they're only available when instantiating the class (I think). What I don't understand really, is why it ws working before..

**Pull request merged**
